### PR TITLE
Add ability to replace a cart item

### PR DESCRIPTION
**Pull request by robinandeer**
## Describe your changes

Allow users to replace an entry in the cart.

Pick up the offer ID from the URL and use it to replace the entry in the cart.

Also use it to prefill some data in the form.

## Justify why they are needed

To be able to add an "edit" button in the cart.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code

### CODE REVIEW

File: apps/store/src/components/CartPage/useEditProductOffer.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  const apolloClient = useApolloClient()
+  const products = useProductMetadata()
+  const router = useRouter()
+
+  const editProductOffer = async (params: Params) => {
+    setState('loading')
+    // TODO: Check if product offer is connected to price intent -- if not, create new one
+
+    const product = products?.find((item) => item.name === params.productName)
+    if (!product) {
```
Comment (alebedev) [inline apps/store/src/components/CartPage/useEditProductOffer.ts:30]: Nit: if you throw here instead, control flow would be more robust.  Currently if programmer error happens before `try` on line 47, it won't be handled

File: apps/store/src/components/ProductPage/PriceIntentContext.tsx [hunk 2]
```diff
@@ -40,19 +42,26 @@ const usePriceIntentContextValue = () => {
...
-        )
+        if (prev) {
+          const matchingOffer = data.priceIntent.offers.find((item) => compareOffer(item, prev))
+          if (matchingOffer) return matchingOffer
+        }
 
-        if (matchingOffer) return matchingOffer
+        if (entryToReplace) {
+          const matchingReplaceOffer = data.priceIntent.offers.find(
+            (item) =>
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/PriceIntentContext.tsx:59]: Nit: can we use `compareOffer` or similar helper here?

File: apps/store/src/components/ProductPage/ProductPage.tsx [hunk 3]
```diff
@@ -22,3 +24,16 @@ export const ProductPage = ({ story, ...props }: ProductPageProps) => {
...
+  const router = useRouter()
+  const { shopSession } = useShopSession()
+
+  const cartEntryId = router.query.replace as string | undefined
+  const productOffer = useMemo(
+    () => shopSession?.cart.entries.find((item) => item.id === cartEntryId),
+    [shopSession?.cart.entries, cartEntryId],
+  )
+
+  return productOffer ?? null
```
Comment (alebedev) [inline apps/store/src/components/ProductPage/ProductPage.tsx:38]: Let's just return memo result, why convent undefined to nul?

**Review by alebedev — APPROVED** [review]
Good one!  Some comments inside

File: apps/store/src/components/ProductPage/PriceIntentContext.tsx [hunk 4]
```diff
@@ -40,19 +42,24 @@ const usePriceIntentContextValue = () => {
...
+          if (matchingOffer) return matchingOffer
+        }
 
-        if (matchingOffer) return matchingOffer
+        if (entryToReplace) {
+          const matchingReplaceOffer = data.priceIntent.offers.find((item) =>
+            compareOffer(item, entryToReplace),
+          )
+          if (matchingReplaceOffer) return matchingReplaceOffer
+        }
```
Comment (guilhermespopolin) [inline apps/store/src/components/ProductPage/PriceIntentContext.tsx:62]: I guess I didn't get this part: Why the selected offer should be the one corresponding with a cart entry that should be replaced, when we have one? 🤔

**Review by guilhermespopolin — APPROVED** [review]

File: apps/store/src/components/ProductPage/PurchaseForm/useHandleSubmitAddToCart.ts [hunk 5]
```diff
@@ -35,7 +40,11 @@ export const useHandleSubmitAddToCart = ({ cartId, onSuccess }: Params) => {
     const productOfferId = getOrThrowFormValue(formData, FormElement.ProductOfferId)
     const nextUrl = event.nativeEvent.submitter?.getAttribute('value') ?? undefined
 
-    addEntry({
+    if (entryToReplace) {
```
Comment (robinandeer) [inline apps/store/src/components/ProductPage/PurchaseForm/useHandleSubmitAddToCart.ts:43]: Do we need to compare this with the entry we are adding?

**Pull request merged**